bus-mapping/src/circuit_input_builder/execution.rs

```diff
@@ -472,6 +472,324 @@ impl CopyEvent {
     }
 }
 
+/// Defines a builder to construct a copy event.
+///
+/// ```markdown
+///     │◄──read_offset──►│
+///     ├─────────────────┼───────────────┬──────┐
+///     │                 │ Source  Bytes │      │
+///     └─────────────────┼───────┬───────┼──────┘
+///      get_padding      │     mapper    │
+/// ┌─────────▼───────────┼───────▼───────┼─────────┐
+/// │      Padding        │ Copied  Bytes │ Padding │
+/// ├─────────────────────┼───────────────┼─────────┤
+/// │◄────write_offset───►│◄───length────►│         │
+/// │◄─────────────── step_length ─────────────────►│
+/// ```
+pub struct CopyEventStepsBuilder<
```

Comment:
why not as following? generic not necessary IMO.
pub struct CopyEventStepsBuilder
 {
    source: Source,
    read_offset: ReadOffset,
    write_offset: WriteOffset,
    step_length: StepLength,
    length: Length,
    padding_byte_getter: Padding,
    mapper: Mapper,
}

Reply:
this builder pattern provides compile time check of the construction, the `build` method only available when everything is ok.
and it may be able to reduce by using other crate like [derive_builder](https://docs.rs/derive_builder) (not a fan of it, or we can make our own version)